File: modules/noorlife-text-recognition/android/src/main/java/com/noorlife/textrecognition/NoorLifeTextRecognitionModule.kt
package com.noorlife.textrecognition

import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Android Latin text recognition, from the bundled ML Kit model — issue #101.
 *
 * ── One recogniser, chosen at build time ────────────────────────────────────
 * `TextRecognizerOptions.DEFAULT_OPTIONS` is the Latin recogniser, and it is the only one this
 * module's `build.gradle` declares. There is no script parameter: a script the caller can choose is
 * a setting that has to be explained, stored and migrated, and it would drag the other four models
 * back into the APK to make the choice meaningful.
 *
 * ── What is refused, and what is never said ────────────────────────────────
 * Anything that is not a local `file://` path is refused before an image is opened. ML Kit's own
 * `InputImage` will read a content URI and the community module it replaces would fetch an `http`
 * URL outright; a workflow that promises a receipt never leaves the device must not have a path that
 * pulls one onto it either.
 *
 * Every rejection carries a fixed message. The URI is app-owned, but the surrounding log is not, and
 * an ML Kit failure message is built from the image it failed on — so neither the path nor the
 * vendor's text is passed back. Nothing here logs.
 */
class NoorLifeTextRecognitionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NoorLifeTextRecognition")

    AsyncFunction("recognizeLatin") { uri: String, promise: Promise ->
      val context = appContext.reactContext
      if (context == null) {
        promise.reject(CodedException("ERR_TEXT_RECOGNITION", "No application context.", null))
        return@AsyncFunction
      }

      if (!uri.trim().startsWith("file:///")) {
        promise.reject(
          CodedException("ERR_TEXT_RECOGNITION_NOT_LOCAL", "Only local files are read.", null),
        )
        return@AsyncFunction
      }

      try {
        val image = InputImage.fromFilePath(context, Uri.parse(uri.trim()))
        TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
          .process(image)
          .addOnSuccessListener { result -> promise.resolve(mapOf("text" to result.text)) }
          .addOnFailureListener {
            promise.reject(
              CodedException("ERR_TEXT_RECOGNITION", "That image could not be read.", null),
            )
          }
      } catch (_: Throwable) {
        promise.reject(
          CodedException("ERR_TEXT_RECOGNITION", "That image could not be read.", null),
        )
      }
    }
  }
}

File: modules/noorlife-text-recognition/ios/NoorLifeTextRecognitionModule.swift
import ExpoModulesCore
import ImageIO
import Vision

/**
 iOS Latin text recognition, from Apple Vision — issue #101.

 ── Why Vision and not ML Kit ───────────────────────────────────────────────
 Vision is part of iOS. It vendors nothing, bundles no model and downloads nothing, so the iOS side
 of Receipts carries no third-party OCR SDK at all — which is both the packaging fix this replaces
 the community module for, and a smaller privacy surface: there is no Google component on this
 platform to disclose.

 The community module could not be configured down to one script on iOS. Its podspec declares five
 GoogleMLKit pods, and CocoaPods offers a consumer no way to drop a dependency a podspec declares.

 ── What is refused, and what is never said ─────────────────────────────────
 Anything that is not a local `file://` path is refused before an image is opened, matching the
 Android implementation and the TypeScript port. Every rejection carries a fixed message: the path is
 app-owned but the surrounding log is not, and a decode failure's own description can name the file.

 ── Why language correction is off ──────────────────────────────────────────
 A receipt is not prose. `usesLanguageCorrection` exists to make recognised words more likely to be
 real words, and on a receipt the interesting tokens are amounts, dates and codes — the exact strings
 a language model is most likely to "improve" into something the receipt does not say.
 */
public class NoorLifeTextRecognitionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NoorLifeTextRecognition")

    AsyncFunction("recognizeLatin") { (uri: String, promise: Promise) in
      let trimmed = uri.trimmingCharacters(in: .whitespacesAndNewlines)

      guard trimmed.hasPrefix("file:///"), let url = URL(string: trimmed) else {
        promise.reject("ERR_TEXT_RECOGNITION_NOT_LOCAL", "Only local files are read.")
        return
      }

      guard
        let source = CGImageSourceCreateWithURL(url as CFURL, nil),
        let image = CGImageSourceCreateImageAtIndex(source, 0, nil)
      else {
        promise.reject("ERR_TEXT_RECOGNITION", "That image could not be read.")
        return
      }

      let request = VNRecognizeTextRequest { request, error in
        if error != nil {
          promise.reject("ERR_TEXT_RECOGNITION", "That image could not be read.")
          return
        }
        let observations = request.results as? [VNRecognizedTextObservation] ?? []
        /*
          One line per observation, joined the way ML Kit joins its own blocks, so the TypeScript
          port receives the same shape from both platforms and the parser has one input format.
        */
        let text = observations
          .compactMap { $0.topCandidates(1).first?.string }
          .joined(separator: "\n")
        promise.resolve(["text": text])
      }
      request.recognitionLevel = .accurate
      request.usesLanguageCorrection = false
      request.recognitionLanguages = ["en-US"]

      DispatchQueue.global(qos: .userInitiated).async {
        do {
          try VNImageRequestHandler(cgImage: image, options: [:]).perform([request])
        } catch {
          promise.reject("ERR_TEXT_RECOGNITION", "That image could not be read.")
        }
      }
    }
  }
}
